packages/react-compose/src/next/README.md

```diff
@@ -0,0 +1,226 @@
+# @fluentui/react-compose
+
+`@fluentui/react-compose` is a library of utilities for composing Fluent UI components.
+
+## A basic component walkthrough
+
+Building a recomposable component requires that we build legos; we put them together, but we let others put them together how they may need. Here's what's needed:
+
+- **State hooks** - all processing of the component should be done within one or more reusable hooks. The hooks should be allowed to manipulate the state in some reusable way, so that recomposing the component can be easily done.
+- **Render function** - a function which takes in final state and returns JSX. (e.g. `renderButton`)
+- **Component factory function** - a create function which returns `{ state, render }` parts for creating the component. This is used to scaffold a new version of the component as needed.
+- **Style hooks** - hooks which can take in styles and provide appropriate classnames to the anatomy of the component.
+
+With these building blocks, you can compose or recompose the component in numerous ways.
+
+### Simple example
+
+A render function:
+
+```jsx
+const renderButton = state => {
+  return <button {...state} />;
+};
+```
+
+A hook which can manipulate the state (defining accessibility and behaviors):
+
+```jsx
+const useButton = state => {
+  // If the button is rendered as something besides a button or anchor, make it focusable.
+  if (state.as !== 'button' && state.as !== 'a') {
+    state.tabIndex = 0;
+  }
+};
+```
+
+A factory function sets up the state and render function:
+
+```jsx
+const createButton = (userProps, ref, defaultProps) => {
+  const state = _.merge({}, defaultProps, userProps);
+
+  // Apply button behaviors.
+  useButton(state);
+
+  return { state, render };
+};
+```
+
+A button can now be easily scaffolded, along with your choice of styling systems:
+
+```jsx
+import * as styles from './Button.scss';
+
+const useStyles = makeStyles(styles);
+
+const Button = React.forwardRef((props, ref) => {
+  const { state, render } = createButton(props, ref);
+
+  // Inject classNames as needed.
+  useStyles(state);
+
+  return render(state);
+});
+```
+
+### Details
+
+#### Creating mutable state with `mergeProps`
+
+In the previous example, `_.merge` was used to deep clone the props into a state object. Creating a single clone and using that to construct state simplifies hook development and usage; rather than trying to re-clone objects unnecessarily on every small mutation, hooks can assume operating against a draft state. This creates more self contained hooks, which can ensure they are used correctly, avoiding accidents like stomping on existing event handlers by blind object assigning the results.
```

Comment:
~~Then we will need a custom utility to deep merge JSX objects (i.e. `children` or slots) because JSX objects are deep nested may be circular: https://github.com/FormidableLabs/react-fast-compare~~